ElasticFrameProtocol.cpp

```diff
@@ -420,119 +426,107 @@ ElasticFrameProtocolReceiver::unpackType3(const uint8_t *pSubPacket, size_t lPac
                 //We got HOL but the next frame has timed out meaning the time out of the bucket + the HOL timeout
                 //We need now need to jump ahead and reset the mNextExpectedFrameNumber
                 //Assemble all data for delivery and reset the HOL pointer.
 
-                if (rBucket->mDeliveryOrder < mNextExpectedFrameNumber) {
-                    //Remove the data since we dont want to deliver OOO
-                    mBucketMap.erase(rBucket->mDeliveryOrder);
-                    rBucket->mBucketData = nullptr;
-                    rBucket->mActive = false;
+                rBucket.second->mBucketData->mDataContent = rBucket.second->mDataContent;
+                rBucket.second->mBucketData->mBroken =
+                        rBucket.second->mFragmentCounter != rBucket.second->mOfFragmentNo;
+                rBucket.second->mBucketData->mPts = rBucket.second->mPts;
+                rBucket.second->mBucketData->mDts = rBucket.second->mDts;
+                rBucket.second->mBucketData->mCode = rBucket.second->mCode;
+                rBucket.second->mBucketData->mStreamID = rBucket.second->mStream;
+                rBucket.second->mBucketData->mSource = rBucket.second->mSource;
+                rBucket.second->mBucketData->mFlags = rBucket.second->mFlags;
+                if (rReceiveFunction) {
+                    rReceiveFunction(rBucket.second->mBucketData, mCTX ? mCTX.get() : nullptr);
                 } else {
-                    rBucket->mBucketData->mDataContent = rBucket->mDataContent;
-                    rBucket->mBucketData->mBroken = rBucket->mFragmentCounter != rBucket->mOfFragmentNo;
-                    rBucket->mBucketData->mPts = rBucket->mPts;
-                    rBucket->mBucketData->mDts = rBucket->mDts;
-                    rBucket->mBucketData->mCode = rBucket->mCode;
-                    rBucket->mBucketData->mStreamID = rBucket->mStream;
-                    rBucket->mBucketData->mSource = rBucket->mSource;
-                    rBucket->mBucketData->mFlags = rBucket->mFlags;
-                    if (rReceiveFunction) {
-                        rReceiveFunction(rBucket->mBucketData, mCTX ? mCTX.get() : nullptr);
-                    } else {
-                        receiveCallback(rBucket->mBucketData, mCTX ? mCTX.get() : nullptr);
-                    }
-                    mBucketMap.erase(rBucket->mDeliveryOrder); //We delivered let's collect the garbage
-                    rBucket->mActive = false; //Inactivate the bucket
-                    rBucket->mBucketData = nullptr; //Release the data
-                    mNextExpectedFrameNumber = rBucket->mDeliveryOrder + 1;
+                    receiveCallback(rBucket.second->mBucketData, mCTX ? mCTX.get() : nullptr);
                 }
+                lDeleteList.emplace_back(rBucket.first);
+                mNextExpectedFrameNumber = rBucket.second->mDeliveryOrder + 1;
```

Comment:
Here we know that we have failed to deliver `mNextExpectedFrameNumber - rBucket.second->mDeliveryOrder` number of SuperFrames, from which we have not received any fragments at all. I don't know if is a good idea to call the callbacks also for these frames, with the broken flag set to true to allow the application layer keeping score of lost super frames? Maybe not, since we cannot do that for non-HOL mode, plus that if we lose many frames over a couple of seconds, the callback will be called many times in a row, which might not be very efficient. Maybe a better solution is to provide the super frame counter with the callback? In case of HOL mode the application can easily calculate the diff to the latest received super frame to check if any frames has gone completely missing, and it would be possible to keep track of missing super frames in non-HOL mode as well by keeping track of past received frames.

Anyway, that is for another PR.

Reply:
Yes makes sense!  Will think about a possible solution.